From ed601a265951a93031dd314ba0ba36c01deeed32 Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Sun, 15 Jun 2025 13:13:26 +0300
Subject: [PATCH] platform/mellanox: nvsw-host-spc6: Add system control and
 monitoring driver for Nvidia SPC6 switches

The driver allows system control and monitoring of Nvidia SPC6 based
switches through CPLD/FPGA devices from the host CPU.

This control includes:
- Handling of hotplug events.
- Various resets handling through sysfs attributes.
- Reset cause reports through sysfs attributes.
- Mux attributes for hw components selection.
- General purpose registers attributes.
- LED control.
- FAN control.

Add the nvsw core, which implements the register map and the device
creation flow shared by all the nvsw host platform drivers, along with
the register definitions used by them.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/platform/mellanox/Kconfig          |  10 +
 drivers/platform/mellanox/Makefile         |   1 +
 drivers/platform/mellanox/nvsw-core.c      | 693 ++++++++++++++++
 drivers/platform/mellanox/nvsw-host-spc6.c | 896 +++++++++++++++++++++
 drivers/platform/mellanox/nvsw.h           | 300 +++++++
 5 files changed, 1900 insertions(+)
 create mode 100644 drivers/platform/mellanox/nvsw-core.c
 create mode 100644 drivers/platform/mellanox/nvsw-host-spc6.c
 create mode 100644 drivers/platform/mellanox/nvsw.h

diff --git a/drivers/platform/mellanox/Kconfig b/drivers/platform/mellanox/Kconfig
index ebb76ab..99b95e9 100644
--- a/drivers/platform/mellanox/Kconfig
+++ b/drivers/platform/mellanox/Kconfig
@@ -113,6 +113,16 @@ config MLXBF_PMC
 	  to performance monitoring counters within various blocks in the
 	  Mellanox BlueField SoC via a sysfs interface.
 
+config NVSW_HOST_SPC6
+	tristate "Nvidia SPC6 host CPLD/FPGA Hardware Control and Monitoring"
+	depends on HWMON && I2C
+	select REGMAP_I2C
+	help
+	  Say Y here to include support for the FPGA/CPLD logic by host CPU
+	  equipped on Nvidia SPC6 ethernet switches.
+	  This driver can also be built as a module. If so the module
+	  will be called nvsw-host-spc6.
+
 config NVSW_SN2201
 	tristate "Nvidia SN2201 platform driver support"
 	depends on HWMON && I2C
diff --git a/drivers/platform/mellanox/Makefile b/drivers/platform/mellanox/Makefile
index 5a8fc58..2c06058 100644
--- a/drivers/platform/mellanox/Makefile
+++ b/drivers/platform/mellanox/Makefile
@@ -11,5 +11,6 @@ obj-$(CONFIG_MLXBF_PTM)	+= mlxbf-ptm.o
 obj-$(CONFIG_MLXREG_HOTPLUG)	+= mlxreg-hotplug.o
 obj-$(CONFIG_MLXREG_IO) += mlxreg-io.o
 obj-$(CONFIG_MLXREG_LC) += mlxreg-lc.o
+obj-$(CONFIG_NVSW_HOST_SPC6) += nvsw-host-spc6.o nvsw-core.o
 obj-$(CONFIG_NVSW_SN2201) += nvsw-sn2201.o
 obj-$(CONFIG_MLXREG_DPU)	+= mlxreg-dpu.o
diff --git a/drivers/platform/mellanox/nvsw-core.c b/drivers/platform/mellanox/nvsw-core.c
new file mode 100644
index 0000000..e309378
--- /dev/null
+++ b/drivers/platform/mellanox/nvsw-core.c
@@ -0,0 +1,693 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Nvidia switch platform driver
+ *
+ * Copyright (C) 2025 Nvidia Technologies Ltd.
+ */
+
+#include <linux/device.h>
+#include <linux/gpio.h>
+#include <linux/i2c.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+#include <linux/of_irq.h>
+#include <linux/platform_data/mlxreg.h>
+#include <linux/platform_device.h>
+#include <linux/regmap.h>
+#include "nvsw.h"
+
+static bool nvsw_writeable_reg(struct device *dev, unsigned int reg)
+{
+	switch (reg) {
+	case NVSW_REG_PG1_EVENT_OFFSET:
+	case NVSW_REG_PG1_MASK_OFFSET:
+	case NVSW_REG_PG2_EVENT_OFFSET:
+	case NVSW_REG_PG2_MASK_OFFSET:
+	case NVSW_REG_PG3_EVENT_OFFSET:
+	case NVSW_REG_PG3_MASK_OFFSET:
+	case NVSW_REG_PG4_EVENT_OFFSET:
+	case NVSW_REG_PG4_MASK_OFFSET:
+	case NVSW_REG_RESET_GP1_OFFSET:
+	case NVSW_REG_FIELD_UPGRADE:
+	case NVSW_REG_GP0_OFFSET:
+	case NVSW_REG_GP1_OFFSET:
+	case NVSW_REG_GP7_OFFSET:
+	case NVSW_REG_PWM_CONTROL_OFFSET:
+	case NVSW_REG_RESET_GP2_OFFSET:
+	case NVSW_REG_GP4_OFFSET:
+	case NVSW_REG_GP5_OFFSET:
+	case NVSW_REG_GP6_OFFSET:
+	case NVSW_REG_LED1_OFFSET:
+	case NVSW_REG_LED5_OFFSET:
+	case NVSW_REG_LED6_OFFSET:
+	case NVSW_REG_LED7_OFFSET:
+	case NVSW_REG_AGGRCO_MASK_OFFSET:
+	case NVSW_REG_HEALTH_EVENT_OFFSET:
+	case NVSW_REG_HEALTH_MASK_OFFSET:
+	case NVSW_REG_AGGR_MASK_OFFSET:
+	case NVSW_REG_FU_CAP_OFFSET:
+	case NVSW_REG_BRD4_EVENT_OFFSET:
+	case NVSW_REG_BRD4_MASK_OFFSET:
+	case NVSW_REG_AGGRLO_MASK_OFFSET:
+	case NVSW_REG_BRD1_EVENT_OFFSET:
+	case NVSW_REG_BRD1_MASK_OFFSET:
+	case NVSW_REG_ASIC1_EVENT_OFFSET:
+	case NVSW_REG_ASIC1_MASK_OFFSET:
+	case NVSW_REG_ASIC2_EVENT_OFFSET:
+	case NVSW_REG_ASIC2_MASK_OFFSET:
+	case NVSW_REG_ASIC3_EVENT_OFFSET:
+	case NVSW_REG_ASIC3_MASK_OFFSET:
+	case NVSW_REG_VR1_ALERT_EVENT_OFFSET:
+	case NVSW_REG_VR1_ALERT_MASK_OFFSET:
+	case NVSW_REG_VR2_ALERT_EVENT_OFFSET:
+	case NVSW_REG_VR2_ALERT_MASK_OFFSET:
+	case NVSW_REG_PS_DC_OK_EVENT_OFFSET:
+	case NVSW_REG_PS_DC_OK_MASK_OFFSET:
+	case NVSW_REG_FAN_EVENT_OFFSET:
+	case NVSW_REG_FAN_MASK_OFFSET:
+	case NVSW_REG_PWRB_EVENT_OFFSET:
+	case NVSW_REG_PWRB_MASK_OFFSET:
+	case NVSW_REG_EROT_EVENT_OFFSET:
+	case NVSW_REG_EROT_MASK_OFFSET:
+	case NVSW_REG_EROT_ERR_EVENT_OFFSET:
+	case NVSW_REG_EROT_ERR_MASK_OFFSET:
+	case NVSW_REG_FRU1_EVENT_OFFSET:
+	case NVSW_REG_FRU1_MASK_OFFSET:
+	case NVSW_REG_LEAK_EVENT_OFFSET:
+	case NVSW_REG_LEAK_MASK_OFFSET:
+	case NVSW_REG_SPI_CHNL_SELECT:
+	case NVSW_REG_WD2_TMR_OFFSET:
+	case NVSW_REG_WD2_TLEFT_OFFSET:
+	case NVSW_REG_WD2_ACT_OFFSET:
+	case NVSW_REG_WD3_TMR_OFFSET:
+	case NVSW_REG_WD3_TLEFT_OFFSET:
+	case NVSW_REG_WD3_ACT_OFFSET:
+	case NVSW_REG_PWM1_OFFSET:
+	case NVSW_REG_MUX0_OFFSET:
+	case NVSW_REG_MUX1_OFFSET:
+	case NVSW_REG_MUX2_OFFSET:
+		return true;
+	}
+	return false;
+}
+
+static bool nvsw_readable_reg(struct device *dev, unsigned int reg)
+{
+	switch (reg) {
+	case NVSW_REG_CPLD1_VER_OFFSET:
+	case NVSW_REG_CPLD1_PN_OFFSET:
+	case NVSW_REG_CPLD1_PN1_OFFSET:
+	case NVSW_REG_CPLD2_VER_OFFSET:
+	case NVSW_REG_CPLD2_PN_OFFSET:
+	case NVSW_REG_CPLD2_PN1_OFFSET:
+	case NVSW_REG_CPLD3_VER_OFFSET:
+	case NVSW_REG_CPLD3_PN_OFFSET:
+	case NVSW_REG_CPLD3_PN1_OFFSET:
+	case NVSW_REG_CPLD4_VER_OFFSET:
+	case NVSW_REG_CPLD4_PN_OFFSET:
+	case NVSW_REG_CPLD4_PN1_OFFSET:
+	case NVSW_REG_CPLD7_VER_OFFSET:
+	case NVSW_REG_PG1_OFFSET:
+	case NVSW_REG_PG1_EVENT_OFFSET:
+	case NVSW_REG_PG1_MASK_OFFSET:
+	case NVSW_REG_PG2_OFFSET:
+	case NVSW_REG_PG2_EVENT_OFFSET:
+	case NVSW_REG_PG2_MASK_OFFSET:
+	case NVSW_REG_PG3_OFFSET:
+	case NVSW_REG_PG3_EVENT_OFFSET:
+	case NVSW_REG_PG3_MASK_OFFSET:
+	case NVSW_REG_PG4_OFFSET:
+	case NVSW_REG_PG4_EVENT_OFFSET:
+	case NVSW_REG_PG4_MASK_OFFSET:
+	case NVSW_REG_CPLD6_VER_OFFSET:
+	case NVSW_REG_CPLD6_PN_OFFSET:
+	case NVSW_REG_CPLD6_PN1_OFFSET:
+	case NVSW_REG_CPLD9_VER_OFFSET:
+	case NVSW_REG_CPLD10_PN_OFFSET:
+	case NVSW_REG_CPLD10_PN1_OFFSET:
+	case NVSW_REG_RESET_GP1_OFFSET:
+	case NVSW_REG_FIELD_UPGRADE:
+	case NVSW_REG_SAFE_BIOS_OFFSET:
+	case NVSW_REG_RESET_CAUSE_OFFSET:
+	case NVSW_REG_RESET_CAUSE1_OFFSET:
+	case NVSW_REG_RESET_CAUSE2_OFFSET:
+	case NVSW_REG_LED1_OFFSET:
+	case NVSW_REG_LED5_OFFSET:
+	case NVSW_REG_LED6_OFFSET:
+	case NVSW_REG_LED7_OFFSET:
+	case NVSW_REG_CPLD7_PN_OFFSET:
+	case NVSW_REG_CPLD7_PN1_OFFSET:
+	case NVSW_REG_RESET_GP2_OFFSET:
+	case NVSW_REG_GP0_RO_OFFSET:
+	case NVSW_REG_GP1_RO_OFFSET:
+	case NVSW_REG_GP4_RO_OFFSET:
+	case NVSW_REG_GP5_RO_OFFSET:
+	case NVSW_REG_GPCOM0_OFFSET:
+	case NVSW_REG_GP0_OFFSET:
+	case NVSW_REG_GP1_OFFSET:
+	case NVSW_REG_GP7_OFFSET:
+	case NVSW_REG_PWM_CONTROL_OFFSET:
+	case NVSW_REG_GP4_OFFSET:
+	case NVSW_REG_GP5_OFFSET:
+	case NVSW_REG_GP6_OFFSET:
+	case NVSW_REG_AGGRCO_OFFSET:
+	case NVSW_REG_AGGRCO_MASK_OFFSET:
+	case NVSW_REG_HEALTH_OFFSET:
+	case NVSW_REG_HEALTH_EVENT_OFFSET:
+	case NVSW_REG_HEALTH_MASK_OFFSET:
+	case NVSW_REG_AGGR_OFFSET:
+	case NVSW_REG_AGGR_MASK_OFFSET:
+	case NVSW_REG_FU_CAP_OFFSET:
+	case NVSW_REG_BRD4_OFFSET:
+	case NVSW_REG_BRD4_EVENT_OFFSET:
+	case NVSW_REG_BRD4_MASK_OFFSET:
+	case NVSW_REG_AGGRLO_OFFSET:
+	case NVSW_REG_AGGRLO_MASK_OFFSET:
+	case NVSW_REG_BRD1_OFFSET:
+	case NVSW_REG_BRD1_EVENT_OFFSET:
+	case NVSW_REG_BRD1_MASK_OFFSET:
+	case NVSW_REG_ASIC1_HEALTH_OFFSET:
+	case NVSW_REG_ASIC1_EVENT_OFFSET:
+	case NVSW_REG_ASIC1_MASK_OFFSET:
+	case NVSW_REG_ASIC2_HEALTH_OFFSET:
+	case NVSW_REG_ASIC2_EVENT_OFFSET:
+	case NVSW_REG_ASIC2_MASK_OFFSET:
+	case NVSW_REG_ASIC3_HEALTH_OFFSET:
+	case NVSW_REG_ASIC3_EVENT_OFFSET:
+	case NVSW_REG_ASIC3_MASK_OFFSET:
+	case NVSW_REG_VR1_ALERT_OFFSET:
+	case NVSW_REG_VR1_ALERT_EVENT_OFFSET:
+	case NVSW_REG_VR1_ALERT_MASK_OFFSET:
+	case NVSW_REG_VR2_ALERT_OFFSET:
+	case NVSW_REG_VR2_ALERT_EVENT_OFFSET:
+	case NVSW_REG_VR2_ALERT_MASK_OFFSET:
+	case NVSW_REG_PS_DC_OK_OFFSET:
+	case NVSW_REG_PS_DC_OK_EVENT_OFFSET:
+	case NVSW_REG_PS_DC_OK_MASK_OFFSET:
+	case NVSW_REG_CPLD8_PN_OFFSET:
+	case NVSW_REG_CPLD8_PN1_OFFSET:
+	case NVSW_REG_CPLD8_VER_OFFSET:
+	case NVSW_REG_FAN_OFFSET:
+	case NVSW_REG_FAN_EVENT_OFFSET:
+	case NVSW_REG_FAN_MASK_OFFSET:
+	case NVSW_REG_CPLD5_VER_OFFSET:
+	case NVSW_REG_CPLD5_PN_OFFSET:
+	case NVSW_REG_CPLD5_PN1_OFFSET:
+	case NVSW_REG_EROT_OFFSET:
+	case NVSW_REG_EROT_EVENT_OFFSET:
+	case NVSW_REG_EROT_MASK_OFFSET:
+	case NVSW_REG_EROT_ERR_OFFSET:
+	case NVSW_REG_EROT_ERR_EVENT_OFFSET:
+	case NVSW_REG_EROT_ERR_MASK_OFFSET:
+	case NVSW_REG_PWRB_OFFSET:
+	case NVSW_REG_PWRB_EVENT_OFFSET:
+	case NVSW_REG_PWRB_MASK_OFFSET:
+	case NVSW_REG_FRU1_OFFSET:
+	case NVSW_REG_FRU1_EVENT_OFFSET:
+	case NVSW_REG_FRU1_MASK_OFFSET:
+	case NVSW_REG_LEAK_OFFSET:
+	case NVSW_REG_LEAK_EVENT_OFFSET:
+	case NVSW_REG_LEAK_MASK_OFFSET:
+	case NVSW_REG_CPLD1_MVER_OFFSET:
+	case NVSW_REG_CPLD2_MVER_OFFSET:
+	case NVSW_REG_CPLD3_MVER_OFFSET:
+	case NVSW_REG_CPLD4_MVER_OFFSET:
+	case NVSW_REG_PWM1_OFFSET:
+	case NVSW_REG_TACHO1_OFFSET:
+	case NVSW_REG_TACHO2_OFFSET:
+	case NVSW_REG_TACHO3_OFFSET:
+	case NVSW_REG_TACHO4_OFFSET:
+	case NVSW_REG_TACHO5_OFFSET:
+	case NVSW_REG_TACHO6_OFFSET:
+	case NVSW_REG_TACHO7_OFFSET:
+	case NVSW_REG_TACHO8_OFFSET:
+	case NVSW_REG_TACHO9_OFFSET:
+	case NVSW_REG_TACHO10_OFFSET:
+	case NVSW_REG_TACHO11_OFFSET:
+	case NVSW_REG_TACHO12_OFFSET:
+	case NVSW_REG_FAN_CAP1_OFFSET:
+	case NVSW_REG_FAN_DRW_CAP_OFFSET:
+	case NVSW_REG_TACHO_SPEED_OFFSET:
+	case NVSW_REG_CONFIG1_OFFSET:
+	case NVSW_REG_CONFIG2_OFFSET:
+	case NVSW_REG_CONFIG3_OFFSET:
+	case NVSW_REG_SPI_CHNL_SELECT:
+	case NVSW_REG_CPLD5_MVER_OFFSET:
+	case NVSW_REG_CPLD9_PN_OFFSET:
+	case NVSW_REG_CPLD9_PN1_OFFSET:
+	case NVSW_REG_CPLD10_VER_OFFSET:
+	case NVSW_REG_WD2_TMR_OFFSET:
+	case NVSW_REG_WD2_TLEFT_OFFSET:
+	case NVSW_REG_WD2_ACT_OFFSET:
+	case NVSW_REG_WD3_TMR_OFFSET:
+	case NVSW_REG_WD3_TLEFT_OFFSET:
+	case NVSW_REG_WD3_ACT_OFFSET:
+	case NVSW_REG_CPLD7_MVER_OFFSET:
+	case NVSW_REG_CPLD8_MVER_OFFSET:
+	case NVSW_REG_CPLD9_MVER_OFFSET:
+	case NVSW_REG_CPLD10_MVER_OFFSET:
+	case NVSW_REG_CPLD6_MVER_OFFSET:
+	case NVSW_REG_MUX0_OFFSET:
+	case NVSW_REG_MUX1_OFFSET:
+	case NVSW_REG_MUX2_OFFSET:
+	case NVSW_REG_UFM_VERSION_OFFSET:
+	case NVSW_REG_PSU_I2C_CAP_OFFSET:
+		return true;
+	}
+	return false;
+}
+
+static bool nvsw_volatile_reg(struct device *dev, unsigned int reg)
+{
+	switch (reg) {
+	case NVSW_REG_CPLD1_VER_OFFSET:
+	case NVSW_REG_CPLD1_PN_OFFSET:
+	case NVSW_REG_CPLD1_PN1_OFFSET:
+	case NVSW_REG_CPLD2_VER_OFFSET:
+	case NVSW_REG_CPLD2_PN_OFFSET:
+	case NVSW_REG_CPLD2_PN1_OFFSET:
+	case NVSW_REG_CPLD3_VER_OFFSET:
+	case NVSW_REG_CPLD3_PN_OFFSET:
+	case NVSW_REG_CPLD3_PN1_OFFSET:
+	case NVSW_REG_CPLD4_VER_OFFSET:
+	case NVSW_REG_CPLD4_PN_OFFSET:
+	case NVSW_REG_CPLD4_PN1_OFFSET:
+	case NVSW_REG_CPLD7_VER_OFFSET:
+	case NVSW_REG_PG1_OFFSET:
+	case NVSW_REG_PG1_EVENT_OFFSET:
+	case NVSW_REG_PG1_MASK_OFFSET:
+	case NVSW_REG_PG2_OFFSET:
+	case NVSW_REG_PG2_EVENT_OFFSET:
+	case NVSW_REG_PG2_MASK_OFFSET:
+	case NVSW_REG_PG3_OFFSET:
+	case NVSW_REG_PG3_EVENT_OFFSET:
+	case NVSW_REG_PG3_MASK_OFFSET:
+	case NVSW_REG_PG4_OFFSET:
+	case NVSW_REG_PG4_EVENT_OFFSET:
+	case NVSW_REG_PG4_MASK_OFFSET:
+	case NVSW_REG_CPLD6_VER_OFFSET:
+	case NVSW_REG_CPLD6_PN_OFFSET:
+	case NVSW_REG_CPLD6_PN1_OFFSET:
+	case NVSW_REG_CPLD9_VER_OFFSET:
+	case NVSW_REG_CPLD10_PN_OFFSET:
+	case NVSW_REG_CPLD10_PN1_OFFSET:
+	case NVSW_REG_RESET_GP1_OFFSET:
+	case NVSW_REG_FIELD_UPGRADE:
+	case NVSW_REG_SAFE_BIOS_OFFSET:
+	case NVSW_REG_RESET_CAUSE_OFFSET:
+	case NVSW_REG_RESET_CAUSE1_OFFSET:
+	case NVSW_REG_RESET_CAUSE2_OFFSET:
+	case NVSW_REG_LED1_OFFSET:
+	case NVSW_REG_LED5_OFFSET:
+	case NVSW_REG_LED6_OFFSET:
+	case NVSW_REG_LED7_OFFSET:
+	case NVSW_REG_CPLD7_PN_OFFSET:
+	case NVSW_REG_CPLD7_PN1_OFFSET:
+	case NVSW_REG_RESET_GP2_OFFSET:
+	case NVSW_REG_GP0_RO_OFFSET:
+	case NVSW_REG_GP1_RO_OFFSET:
+	case NVSW_REG_GPCOM0_OFFSET:
+	case NVSW_REG_GP0_OFFSET:
+	case NVSW_REG_GP1_OFFSET:
+	case NVSW_REG_GP7_OFFSET:
+	case NVSW_REG_PWM_CONTROL_OFFSET:
+	case NVSW_REG_GP4_OFFSET:
+	case NVSW_REG_GP5_OFFSET:
+	case NVSW_REG_GP6_OFFSET:
+	case NVSW_REG_AGGRCO_OFFSET:
+	case NVSW_REG_AGGRCO_MASK_OFFSET:
+	case NVSW_REG_HEALTH_OFFSET:
+	case NVSW_REG_HEALTH_EVENT_OFFSET:
+	case NVSW_REG_HEALTH_MASK_OFFSET:
+	case NVSW_REG_AGGR_OFFSET:
+	case NVSW_REG_AGGR_MASK_OFFSET:
+	case NVSW_REG_FU_CAP_OFFSET:
+	case NVSW_REG_BRD4_OFFSET:
+	case NVSW_REG_BRD4_EVENT_OFFSET:
+	case NVSW_REG_BRD4_MASK_OFFSET:
+	case NVSW_REG_AGGRLO_OFFSET:
+	case NVSW_REG_AGGRLO_MASK_OFFSET:
+	case NVSW_REG_BRD1_OFFSET:
+	case NVSW_REG_BRD1_EVENT_OFFSET:
+	case NVSW_REG_BRD1_MASK_OFFSET:
+	case NVSW_REG_ASIC1_HEALTH_OFFSET:
+	case NVSW_REG_ASIC1_EVENT_OFFSET:
+	case NVSW_REG_ASIC1_MASK_OFFSET:
+	case NVSW_REG_ASIC2_HEALTH_OFFSET:
+	case NVSW_REG_ASIC2_EVENT_OFFSET:
+	case NVSW_REG_ASIC2_MASK_OFFSET:
+	case NVSW_REG_ASIC3_HEALTH_OFFSET:
+	case NVSW_REG_ASIC3_EVENT_OFFSET:
+	case NVSW_REG_ASIC3_MASK_OFFSET:
+	case NVSW_REG_VR1_ALERT_OFFSET:
+	case NVSW_REG_VR1_ALERT_EVENT_OFFSET:
+	case NVSW_REG_VR1_ALERT_MASK_OFFSET:
+	case NVSW_REG_VR2_ALERT_OFFSET:
+	case NVSW_REG_VR2_ALERT_EVENT_OFFSET:
+	case NVSW_REG_VR2_ALERT_MASK_OFFSET:
+	case NVSW_REG_PS_DC_OK_OFFSET:
+	case NVSW_REG_PS_DC_OK_EVENT_OFFSET:
+	case NVSW_REG_PS_DC_OK_MASK_OFFSET:
+	case NVSW_REG_CPLD8_PN_OFFSET:
+	case NVSW_REG_CPLD8_PN1_OFFSET:
+	case NVSW_REG_CPLD8_VER_OFFSET:
+	case NVSW_REG_FAN_OFFSET:
+	case NVSW_REG_FAN_EVENT_OFFSET:
+	case NVSW_REG_FAN_MASK_OFFSET:
+	case NVSW_REG_CPLD5_VER_OFFSET:
+	case NVSW_REG_CPLD5_PN_OFFSET:
+	case NVSW_REG_CPLD5_PN1_OFFSET:
+	case NVSW_REG_EROT_OFFSET:
+	case NVSW_REG_EROT_EVENT_OFFSET:
+	case NVSW_REG_EROT_MASK_OFFSET:
+	case NVSW_REG_EROT_ERR_OFFSET:
+	case NVSW_REG_EROT_ERR_EVENT_OFFSET:
+	case NVSW_REG_EROT_ERR_MASK_OFFSET:
+	case NVSW_REG_PWRB_OFFSET:
+	case NVSW_REG_PWRB_EVENT_OFFSET:
+	case NVSW_REG_PWRB_MASK_OFFSET:
+	case NVSW_REG_FRU1_OFFSET:
+	case NVSW_REG_FRU1_EVENT_OFFSET:
+	case NVSW_REG_FRU1_MASK_OFFSET:
+	case NVSW_REG_LEAK_OFFSET:
+	case NVSW_REG_LEAK_EVENT_OFFSET:
+	case NVSW_REG_LEAK_MASK_OFFSET:
+	case NVSW_REG_GP4_RO_OFFSET:
+	case NVSW_REG_GP5_RO_OFFSET:
+	case NVSW_REG_CPLD1_MVER_OFFSET:
+	case NVSW_REG_CPLD2_MVER_OFFSET:
+	case NVSW_REG_CPLD3_MVER_OFFSET:
+	case NVSW_REG_CPLD4_MVER_OFFSET:
+	case NVSW_REG_PWM1_OFFSET:
+	case NVSW_REG_TACHO1_OFFSET:
+	case NVSW_REG_TACHO2_OFFSET:
+	case NVSW_REG_TACHO3_OFFSET:
+	case NVSW_REG_TACHO4_OFFSET:
+	case NVSW_REG_TACHO5_OFFSET:
+	case NVSW_REG_TACHO6_OFFSET:
+	case NVSW_REG_TACHO7_OFFSET:
+	case NVSW_REG_TACHO8_OFFSET:
+	case NVSW_REG_TACHO9_OFFSET:
+	case NVSW_REG_TACHO10_OFFSET:
+	case NVSW_REG_TACHO11_OFFSET:
+	case NVSW_REG_TACHO12_OFFSET:
+	case NVSW_REG_FAN_CAP1_OFFSET:
+	case NVSW_REG_FAN_DRW_CAP_OFFSET:
+	case NVSW_REG_TACHO_SPEED_OFFSET:
+	case NVSW_REG_CONFIG1_OFFSET:
+	case NVSW_REG_CONFIG2_OFFSET:
+	case NVSW_REG_CONFIG3_OFFSET:
+	case NVSW_REG_SPI_CHNL_SELECT:
+	case NVSW_REG_CPLD5_MVER_OFFSET:
+	case NVSW_REG_CPLD9_PN_OFFSET:
+	case NVSW_REG_CPLD9_PN1_OFFSET:
+	case NVSW_REG_CPLD10_VER_OFFSET:
+	case NVSW_REG_WD2_TMR_OFFSET:
+	case NVSW_REG_WD2_TLEFT_OFFSET:
+	case NVSW_REG_WD2_ACT_OFFSET:
+	case NVSW_REG_CPLD7_MVER_OFFSET:
+	case NVSW_REG_CPLD8_MVER_OFFSET:
+	case NVSW_REG_CPLD9_MVER_OFFSET:
+	case NVSW_REG_CPLD10_MVER_OFFSET:
+	case NVSW_REG_CPLD6_MVER_OFFSET:
+	case NVSW_REG_WD3_TMR_OFFSET:
+	case NVSW_REG_WD3_TLEFT_OFFSET:
+	case NVSW_REG_WD3_ACT_OFFSET:
+	case NVSW_REG_MUX0_OFFSET:
+	case NVSW_REG_MUX1_OFFSET:
+	case NVSW_REG_MUX2_OFFSET:
+	case NVSW_REG_UFM_VERSION_OFFSET:
+	case NVSW_REG_PSU_I2C_CAP_OFFSET:
+		return true;
+	}
+	return false;
+}
+
+/* Configuration for the register map of a device with 2 bytes address space. */
+static const struct reg_default nvsw_core_reg_def[] = {
+	{ NVSW_REG_AGGRCO_MASK_OFFSET, GENMASK(1, 0)},
+	{ NVSW_REG_PWM_CONTROL_OFFSET, 0x00 },
+};
+
+static const struct regmap_config nvsw_regmap_i2c_conf = {
+	.reg_bits = 16,
+	.val_bits = 8,
+	.max_register = NVSW_REG_MAX,
+	.cache_type = REGCACHE_FLAT,
+	.reg_defaults = nvsw_core_reg_def,
+	.num_reg_defaults = ARRAY_SIZE(nvsw_core_reg_def),
+	.writeable_reg = nvsw_writeable_reg,
+	.readable_reg = nvsw_readable_reg,
+	.volatile_reg = nvsw_volatile_reg,
+};
+
+static const struct reg_default nvsw_reg_def_l1[] = {
+	{ NVSW_REG_PWM_CONTROL_OFFSET, 0x00 },
+	{ NVSW_REG_WD2_ACT_OFFSET, 0x00 },
+	{ NVSW_REG_WD3_ACT_OFFSET, 0x00 },
+	{ NVSW_REG_LEAK_MASK_OFFSET, 0x3f },
+};
+
+struct nvsw_io_regmap_context {
+	void __iomem *base;
+};
+
+static struct nvsw_io_regmap_context nvsw_io_regmap_ctx;
+
+static int nvsw_reg_read(void *context, unsigned int reg, unsigned int *val)
+{
+	struct nvsw_io_regmap_context *ctx = context;
+
+	*val = ioread8(ctx->base - NVSW_REG_MIN + reg);
+	return 0;
+}
+
+static int nvsw_reg_write(void *context, unsigned int reg, unsigned int val)
+{
+	struct nvsw_io_regmap_context *ctx = context;
+
+	iowrite8(val, ctx->base - NVSW_REG_MIN + reg);
+	return 0;
+}
+
+static const struct regmap_config nvsw_regmap_conf_l1 = {
+	.reg_bits = 8,
+	.val_bits = 8,
+	.max_register = NVSW_REG_MAX,
+	.cache_type = REGCACHE_FLAT,
+	.writeable_reg = nvsw_writeable_reg,
+	.readable_reg = nvsw_readable_reg,
+	.volatile_reg = nvsw_volatile_reg,
+	.reg_defaults = nvsw_reg_def_l1,
+	.num_reg_defaults = ARRAY_SIZE(nvsw_reg_def_l1),
+	.reg_read = nvsw_reg_read,
+	.reg_write = nvsw_reg_write,
+};
+
+static int nvsw_core_platform_init(struct nvsw_core *nvsw_core)
+{
+	int i, err;
+
+	/* Add registers io access driver. */
+	if (nvsw_core->regio_data) {
+		nvsw_core->regio_data->regmap = nvsw_core->regmap;
+		nvsw_core->regio =
+			platform_device_register_resndata(nvsw_core->dev, "mlxreg-io",
+							  PLATFORM_DEVID_NONE, NULL, 0,
+							  nvsw_core->regio_data,
+							  sizeof(*nvsw_core->regio_data));
+		if (IS_ERR(nvsw_core->regio)) {
+			err = PTR_ERR(nvsw_core->regio);
+			dev_err(nvsw_core->dev, "Failed to register mlxreg-io driver\n");
+			goto fail_platform_io_register;
+		}
+	}
+
+	/* Add FAN driver. */
+	if (nvsw_core->fan_data) {
+		nvsw_core->fan_data->regmap = nvsw_core->regmap;
+		nvsw_core->fan = platform_device_register_resndata(nvsw_core->dev, "mlxreg-fan",
+								   PLATFORM_DEVID_NONE, NULL, 0,
+								   nvsw_core->fan_data,
+								   sizeof(*nvsw_core->fan_data));
+		if (IS_ERR(nvsw_core->fan)) {
+			err = PTR_ERR(nvsw_core->fan);
+			dev_err(nvsw_core->dev, "Failed to register mlxreg-fan driver\n");
+			goto fail_platform_fan_register;
+		}
+	}
+
+	/* Add LED driver. */
+	if (nvsw_core->led_data) {
+		nvsw_core->led_data->regmap = nvsw_core->regmap;
+		nvsw_core->led = platform_device_register_resndata(nvsw_core->dev, "leds-mlxreg",
+								   PLATFORM_DEVID_NONE, NULL, 0,
+								   nvsw_core->led_data,
+								   sizeof(*nvsw_core->led_data));
+		if (IS_ERR(nvsw_core->led)) {
+			err = PTR_ERR(nvsw_core->led);
+			dev_err(nvsw_core->dev, "Failed to register leds-mlxreg driver\n");
+			goto fail_platform_leds_register;
+		}
+	}
+
+	/* Add hotplug driver. */
+	if (nvsw_core->hotplug_data && nvsw_core->np) {
+		nvsw_core->hotplug_data->irq = nvsw_core->client->irq;
+		dev_info(nvsw_core->dev, "irq %d\n", nvsw_core->hotplug_data->irq);
+		nvsw_core->hotplug_data->regmap = nvsw_core->regmap;
+		nvsw_core->hotplug =
+			platform_device_register_resndata(nvsw_core->dev, "mlxreg-hotplug",
+							  PLATFORM_DEVID_NONE, NULL, 0,
+							  nvsw_core->hotplug_data,
+							  sizeof(*nvsw_core->hotplug_data));
+		if (IS_ERR(nvsw_core->hotplug)) {
+			err = PTR_ERR(nvsw_core->hotplug);
+			dev_err(nvsw_core->dev, "Failed to register mlxreg-hotplug driver\n");
+			goto fail_platform_hotplug_register;
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(nvsw_core->wd_data); i++) {
+		if (nvsw_core->wd_data[i]) {
+			nvsw_core->wd_data[i]->regmap = nvsw_core->regmap;
+			nvsw_core->wd[i] =
+				platform_device_register_resndata(nvsw_core->dev, "mlx-wdt", i,
+								  NULL, 0, nvsw_core->wd_data[i],
+								  sizeof(*nvsw_core->wd_data[i]));
+			if (IS_ERR(nvsw_core->wd[i])) {
+				err = PTR_ERR(nvsw_core->wd[i]);
+				dev_err(nvsw_core->dev, "Failed to register mlx-wdt driver\n");
+				goto fail_platform_wd_register;
+			}
+		}
+	}
+
+	return 0;
+
+fail_platform_wd_register:
+	while (i--)
+		platform_device_unregister(nvsw_core->wd[i]);
+	if (nvsw_core->hotplug_data)
+		platform_device_unregister(nvsw_core->hotplug);
+fail_platform_hotplug_register:
+	if (nvsw_core->led_data)
+		platform_device_unregister(nvsw_core->led);
+fail_platform_leds_register:
+	if (nvsw_core->fan_data)
+		platform_device_unregister(nvsw_core->fan);
+fail_platform_fan_register:
+	if (nvsw_core->regio_data)
+		platform_device_unregister(nvsw_core->regio);
+fail_platform_io_register:
+	return err;
+}
+
+static void nvsw_core_platform_exit(struct nvsw_core *nvsw_core)
+{
+	int i;
+
+	for (i = NVSW_WD_MAX - 1; i >= 0; i--)
+		platform_device_unregister(nvsw_core->wd[i]);
+	if (nvsw_core->hotplug_data)
+		platform_device_unregister(nvsw_core->hotplug);
+	if (nvsw_core->led_data)
+		platform_device_unregister(nvsw_core->led);
+	if (nvsw_core->fan_data)
+		platform_device_unregister(nvsw_core->fan);
+	if (nvsw_core->regio_data)
+		platform_device_unregister(nvsw_core->regio);
+}
+
+static int nvsw_core_mux_topology_init(struct nvsw_core *nvsw_core)
+{
+	return nvsw_core->mux_init(nvsw_core);
+}
+
+static void nvsw_core_mux_topology_exit(struct nvsw_core *nvsw_core)
+{
+	nvsw_core->mux_exit(nvsw_core);
+}
+
+static int nvsw_core_regmap_init(struct nvsw_core *nvsw_core)
+{
+	int err;
+
+	switch (nvsw_core->regmap_type) {
+	case REGMAP_I2C:
+		nvsw_core->regmap = devm_regmap_init_i2c(nvsw_core->client, &nvsw_regmap_i2c_conf);
+		break;
+	case REGMAP_IO:
+		if (nvsw_core->port_map) {
+			nvsw_io_regmap_ctx.base = nvsw_core->port_map(nvsw_core);
+			if (!nvsw_io_regmap_ctx.base)
+				return -ENOMEM;
+		}
+		nvsw_core->regmap = devm_regmap_init(nvsw_core->dev, NULL, &nvsw_io_regmap_ctx,
+						     &nvsw_regmap_conf_l1);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	if (IS_ERR(nvsw_core->regmap)) {
+		dev_err(nvsw_core->dev, "Failed to create regmap\n");
+		return PTR_ERR(nvsw_core->regmap);
+	}
+
+	/* Sync registers with hardware. */
+	regcache_mark_dirty(nvsw_core->regmap);
+	err = regcache_sync(nvsw_core->regmap);
+	if (err) {
+		dev_err(nvsw_core->dev, "Failed to sync regmap\n");
+		return err;
+	}
+
+	/* Set registers default values. */
+	if (nvsw_core->set_reg_default) {
+		err = nvsw_core->set_reg_default(nvsw_core->regmap);
+		if (err) {
+			dev_err(nvsw_core->dev, "Failed to set default regmap\n");
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+int nvsw_core_init(struct nvsw_core *nvsw_core)
+{
+	int err;
+
+	err = nvsw_core_regmap_init(nvsw_core);
+	if (err)
+		return err;
+
+	err = nvsw_core_mux_topology_init(nvsw_core);
+	if (err)
+		goto nvsw_core_mux_topology_init_fail;
+
+	err = nvsw_core_platform_init(nvsw_core);
+	if (err)
+		goto nvsw_core_platform_init_fail;
+
+	return 0;
+nvsw_core_platform_init_fail:
+	nvsw_core_mux_topology_exit(nvsw_core);
+nvsw_core_mux_topology_init_fail:
+	return err;
+}
+EXPORT_SYMBOL(nvsw_core_init);
+
+void nvsw_core_exit(struct nvsw_core *nvsw_core)
+{
+	nvsw_core_platform_exit(nvsw_core);
+	nvsw_core_mux_topology_exit(nvsw_core);
+	if (nvsw_core->set_reg_default)
+		nvsw_core->set_reg_default(nvsw_core->regmap);
+}
+EXPORT_SYMBOL(nvsw_core_exit);
+
+MODULE_AUTHOR("Vadim Pasternak <vadimp@mellanox.com>");
+MODULE_DESCRIPTION("Nvidia platform driver");
+MODULE_LICENSE("Dual BSD/GPL");
diff --git a/drivers/platform/mellanox/nvsw-host-spc6.c b/drivers/platform/mellanox/nvsw-host-spc6.c
new file mode 100644
index 0000000..f59f452
--- /dev/null
+++ b/drivers/platform/mellanox/nvsw-host-spc6.c
@@ -0,0 +1,896 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Nvidia SPC6 host platform driver
+ *
+ * Copyright (C) 2025 Nvidia Technologies Ltd.
+ */
+
+#include <linux/device.h>
+#include <linux/dmi.h>
+#include <linux/i2c.h>
+#include <linux/i2c-mux.h>
+#include <linux/io.h>
+#include <linux/module.h>
+#include <linux/regmap.h>
+#include <linux/platform_device.h>
+#include <linux/platform_data/mlxcpld.h>
+#include <linux/platform_data/mlxreg.h>
+
+#include "nvsw.h"
+
+#define NVSW_HOST_DEVICE_NAME	"mlxplat"
+
+/* Bus number of mux parent device */
+#define NVSW_I2C_MUX_PARENT_BUS_NR	2
+
+/* Bus number of the last host I2C bus */
+#define NVSW_I2C_LAST_HOST_BUS_NR	3
+
+/* Platform mux channels */
+static int nvsw_host_spc6_mux_channels[] = {
+	1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17,
+	18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
+	33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
+	48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62,
+	63, 64, 65, 66, 67,
+};
+
+/* Platform mux data */
+static struct mlxcpld_mux_plat_data nvsw_host_spc6_mux_data[] = {
+	{
+		.chan_ids = nvsw_host_spc6_mux_channels,
+		.num_adaps = ARRAY_SIZE(nvsw_host_spc6_mux_channels),
+		.sel_reg_addr = NVSW_REG_MUX1_OFFSET,
+		.reg_size = 2,
+	},
+};
+
+/* Platform mux board info */
+static struct i2c_board_info nvsw_host_spc6_mux_brdinfo = {
+	I2C_BOARD_INFO("i2c-mux-mlxcpld", 0x33),
+};
+
+static struct i2c_adapter *nvsw_host_mux_i2c_adapter;
+static struct i2c_client *nvsw_host_mux_i2c_client;
+static struct mlxcpld_mux_plat_data *nvsw_host_mux_data[NVSW_MUX_MAX];
+static struct i2c_board_info *nvsw_host_mux_brdinfo;
+static struct mlxreg_core_platform_data *nvsw_led_data;
+static struct mlxreg_core_platform_data *nvsw_regs_io_data;
+static struct mlxreg_core_platform_data *nvsw_wd_data[NVSW_WD_MAX];
+static int mux_num;
+static enum nvsw_core_hid_type nvsw_host_hid;
+static struct platform_device *nvsw_host_pdev;
+
+/* Platform register access data */
+static struct mlxreg_core_data nvsw_host_spc6_regs_io_data[] = {
+	{
+		.label = "cpld1_version",
+		.reg = NVSW_REG_CPLD1_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld2_version",
+		.reg = NVSW_REG_CPLD2_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld3_version",
+		.reg = NVSW_REG_CPLD3_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld4_version",
+		.reg = NVSW_REG_CPLD4_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld5_version",
+		.reg = NVSW_REG_CPLD5_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld6_version",
+		.reg = NVSW_REG_CPLD6_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld7_version",
+		.reg = NVSW_REG_CPLD7_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld8_version",
+		.reg = NVSW_REG_CPLD8_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld9_version",
+		.reg = NVSW_REG_CPLD9_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld10_version",
+		.reg = NVSW_REG_CPLD10_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld1_pn",
+		.reg = NVSW_REG_CPLD1_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld2_pn",
+		.reg = NVSW_REG_CPLD2_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld3_pn",
+		.reg = NVSW_REG_CPLD3_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld4_pn",
+		.reg = NVSW_REG_CPLD4_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld5_pn",
+		.reg = NVSW_REG_CPLD5_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld6_pn",
+		.reg = NVSW_REG_CPLD6_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld7_pn",
+		.reg = NVSW_REG_CPLD7_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld8_pn",
+		.reg = NVSW_REG_CPLD8_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld9_pn",
+		.reg = NVSW_REG_CPLD9_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld10_pn",
+		.reg = NVSW_REG_CPLD10_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld1_version_min",
+		.reg = NVSW_REG_CPLD1_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld2_version_min",
+		.reg = NVSW_REG_CPLD2_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld3_version_min",
+		.reg = NVSW_REG_CPLD3_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld4_version_min",
+		.reg = NVSW_REG_CPLD4_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld5_version_min",
+		.reg = NVSW_REG_CPLD5_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld6_version_min",
+		.reg = NVSW_REG_CPLD6_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld7_version_min",
+		.reg = NVSW_REG_CPLD7_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld8_version_min",
+		.reg = NVSW_REG_CPLD8_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld9_version_min",
+		.reg = NVSW_REG_CPLD9_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld10_version_min",
+		.reg = NVSW_REG_CPLD10_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "pwr_converter_prog_en",
+		.reg = NVSW_REG_GP7_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0644,
+	},
+	{
+		.label = "graceful_pwr_off",
+		.reg = NVSW_REG_GP7_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+	},
+	{
+		.label = "cpu_mctp_ready",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0644,
+	},
+	{
+		.label = "cpu_shutdown_req",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "vpd_wp",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+		.secured = 1,
+	},
+	{
+		.label = "pcie_asic_reset_dis",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0644,
+	},
+	{
+		.label = "shutdown_unlock",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0644,
+	},
+	{
+		.label = "cpu_power_off_ready",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.mode = 0644,
+	},
+	{
+		.label = "pwr_cycle",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0244,
+	},
+	{
+		.label = "pwr_down",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0200,
+	},
+	{
+		.label = "aux_pwr_cycle",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0200,
+	},
+	{
+		.label = "bmc_to_cpu_ctrl",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0644,
+	},
+	{
+		.label = "uart_sel",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = NVSW_UART_SEL_MASK,
+		.bit = 7,
+		.mode = 0644,
+	},
+	{
+		.label = "leakage1",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage2",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage3",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage4",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage5",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0444,
+	},
+	{
+		.label = "asic_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+	},
+	{
+		.label = "asic1_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0644,
+	},
+	{
+		.label = "asic2_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0644,
+	},
+	{
+		.label = "asic3_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0644,
+	},
+	{
+		.label = "asic4_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0644,
+	},
+	{
+		.label = "reset_long_pb",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_short_pb",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_aux_pwr_or_reload",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_swb_dc_dc_pwr_fail",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_platform",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_swb_wd",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_asic_thermal",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(7),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_soc",
+		.reg = NVSW_REG_RESET_CAUSE1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_system",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_sw_pwr_off",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_cpu_thermal",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_pwr_converter_fail",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_main_51v",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_mgmt_pwr",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(7),
+		.mode = 0444,
+	},
+	{
+		.label = "jtag_cap",
+		.reg = NVSW_REG_FU_CAP_OFFSET,
+		.mask = NVSW_FU_CAP_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "jtag_enable",
+		.reg = NVSW_REG_FIELD_UPGRADE,
+		.mask = GENMASK(1, 0),
+		.bit = 1,
+		.mode = 0644,
+	},
+	{
+		.label = "asic_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "asic2_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC2_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "asic3_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC3_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "asic4_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC4_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "spi_chnl_select",
+		.reg = NVSW_REG_SPI_CHNL_SELECT,
+		.mask = GENMASK(7, 0),
+		.bit = 1,
+		.mode = 0644,
+	},
+	{
+		.label = "pdb1_pwr_status",
+		.reg = NVSW_REG_PS_DC_OK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0644,
+	},
+	{
+		.label = "pdb2_pwr_status",
+		.reg = NVSW_REG_PS_DC_OK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0644,
+	},
+	{
+		.label = "pdb3_pwr_status",
+		.reg = NVSW_REG_PS_DC_OK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0644,
+	},
+	{
+		.label = "pdb4_pwr_status",
+		.reg = NVSW_REG_PS_DC_OK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+	},
+};
+
+static struct mlxreg_core_platform_data nvsw_host_spc6_regs_io = {
+	.data = nvsw_host_spc6_regs_io_data,
+	.counter = ARRAY_SIZE(nvsw_host_spc6_regs_io_data),
+};
+
+/* Platform led data */
+static struct mlxreg_core_data nvsw_host_spc6_lc_led_data[] = {
+	{
+		.label = "status:green",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_LO_NIBBLE_MASK,
+	},
+	{
+		.label = "status:amber",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_LO_NIBBLE_MASK,
+	},
+	{
+		.label = "power:green",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_HI_NIBBLE_MASK,
+	},
+	{
+		.label = "power:amber",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_HI_NIBBLE_MASK,
+	},
+	{
+		.label = "uid:blue",
+		.reg = NVSW_REG_LED5_OFFSET,
+		.mask = NVSW_LED_LO_NIBBLE_MASK,
+	},
+};
+
+static struct mlxreg_core_platform_data nvsw_host_spc6_lc_led = {
+	.data = nvsw_host_spc6_lc_led_data,
+	.counter = ARRAY_SIZE(nvsw_host_spc6_lc_led_data),
+};
+
+/* Watchdog type3 platform data */
+static struct mlxreg_core_data nvsw_host_wd_main_regs_type3[] = {
+	{
+		.label = "action",
+		.reg = NVSW_REG_WD2_ACT_OFFSET,
+		.mask = NVSW_WD_RESET_ACT_MASK,
+		.bit = 0,
+	},
+	{
+		.label = "timeout",
+		.reg = NVSW_REG_WD2_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+		.health_cntr = NVSW_WD3_DFLT_TIMEOUT,
+	},
+	{
+		.label = "timeleft",
+		.reg = NVSW_REG_WD2_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+	},
+	{
+		.label = "ping",
+		.reg = NVSW_REG_WD2_ACT_OFFSET,
+		.mask = NVSW_WD_RESET_ACT_MASK,
+		.bit = 0,
+	},
+	{
+		.label = "reset",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.bit = 6,
+	},
+};
+
+static struct mlxreg_core_data nvsw_host_wd_aux_regs_type3[] = {
+	{
+		.label = "action",
+		.reg = NVSW_REG_WD3_ACT_OFFSET,
+		.mask = NVSW_WD_FAN_ACT_MASK,
+		.bit = 4,
+	},
+	{
+		.label = "timeout",
+		.reg = NVSW_REG_WD3_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+		.health_cntr = NVSW_WD3_DFLT_TIMEOUT,
+	},
+	{
+		.label = "timeleft",
+		.reg = NVSW_REG_WD3_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+	},
+	{
+		.label = "ping",
+		.reg = NVSW_REG_WD3_ACT_OFFSET,
+		.mask = NVSW_WD_FAN_ACT_MASK,
+		.bit = 4,
+	},
+};
+
+static struct mlxreg_core_platform_data nvsw_host_wd_set_type3[] = {
+	{
+		.data = nvsw_host_wd_main_regs_type3,
+		.counter = ARRAY_SIZE(nvsw_host_wd_main_regs_type3),
+		.version = MLX_WDT_TYPE3,
+		.identity = "mlx-wdt-main",
+	},
+	{
+		.data = nvsw_host_wd_aux_regs_type3,
+		.counter = ARRAY_SIZE(nvsw_host_wd_aux_regs_type3),
+		.version = MLX_WDT_TYPE3,
+		.identity = "mlx-wdt-aux",
+	},
+};
+
+/* Callback is used to indicate that all adapter devices have been created. */
+static int
+nvsw_host_spc6_completion_notify(void *handle, struct i2c_adapter *parent,
+				  struct i2c_adapter *adapters[])
+{
+	return 0;
+}
+
+/* Mux init/exit callbacks. */
+static int nvsw_host_spc6_mux_topology_init(struct nvsw_core *nvsw_core)
+{
+	int i, err;
+
+	/* Create mux infrastructure. */
+	for (i = 0; i < nvsw_core->mux_num; i++) {
+		nvsw_core->mux[i] =
+			platform_device_register_resndata(&nvsw_host_mux_i2c_client->dev,
+							  "i2c-mux-mlxcpld", i, NULL,
+							  0, nvsw_host_mux_data[i],
+							  sizeof(*nvsw_host_mux_data[i]));
+		if (IS_ERR(nvsw_core->mux[i])) {
+			dev_err(nvsw_core->dev, "Failed to create mux infra\n");
+			err = PTR_ERR(nvsw_core->mux[i]);
+			nvsw_core->mux[i] = NULL;
+			goto platform_mux_register_fail;
+		}
+	}
+
+	return 0;
+
+platform_mux_register_fail:
+	while (--i >= 0) {
+		platform_device_unregister(nvsw_core->mux[i]);
+		nvsw_core->mux[i] = NULL;
+	}
+
+	return err;
+}
+
+static void nvsw_host_spc6_mux_topology_exit(struct nvsw_core *nvsw_core)
+{
+	int i;
+
+	for (i = 0; i < nvsw_core->mux_num; i++) {
+		if (nvsw_core->mux[i]) {
+			platform_device_unregister(nvsw_core->mux[i]);
+			nvsw_core->mux[i] = NULL;
+		}
+	}
+}
+
+static int __init nvsw_host_dmi_spc6_switch_matched(const struct dmi_system_id *dmi)
+{
+	int i;
+
+	/* Set system configuration. */
+	nvsw_host_hid = HID193;
+	mux_num = ARRAY_SIZE(nvsw_host_spc6_mux_data);
+	for (i = 0; i < mux_num; i++)
+		nvsw_host_mux_data[i] = &nvsw_host_spc6_mux_data[i];
+	nvsw_host_mux_brdinfo = &nvsw_host_spc6_mux_brdinfo;
+	nvsw_led_data = &nvsw_host_spc6_lc_led;
+	nvsw_regs_io_data = &nvsw_host_spc6_regs_io;
+	for (i = 0; i < ARRAY_SIZE(nvsw_host_wd_set_type3); i++)
+		nvsw_wd_data[i] = &nvsw_host_wd_set_type3[i];
+
+	return 1;
+}
+
+static const struct dmi_system_id nvsw_host_dmi_table[] __initconst = {
+	{
+		.callback = nvsw_host_dmi_spc6_switch_matched,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_NAME, "VMOD0025"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "HI193"),
+		},
+	},
+	{
+		.callback = nvsw_host_dmi_spc6_switch_matched,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_NAME, "VMOD0027"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "HI183"),
+		},
+	},
+	{
+		.callback = nvsw_host_dmi_spc6_switch_matched,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_NAME, "VMOD0027"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "HI187"),
+		},
+	},
+	{
+		.callback = nvsw_host_dmi_spc6_switch_matched,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_NAME, "VMOD0027"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "HI188"),
+		},
+	},
+	{ }
+};
+
+MODULE_DEVICE_TABLE(dmi, nvsw_host_dmi_table);
+
+static int nvsw_host_probe(struct platform_device *pdev)
+{
+	struct i2c_adapter *last_host_adapter;
+	struct nvsw_core *nvsw_core;
+	int i, err;
+
+	/* Ensure the I2C host driver has finished registering
+	 * all 4 buses (0-3) to prevent the mux from stealing bus 3
+	 */
+	last_host_adapter = i2c_get_adapter(NVSW_I2C_LAST_HOST_BUS_NR);
+	if (!last_host_adapter) {
+		dev_info(&pdev->dev,
+			 "I2C adapter for bus %d is not ready, deferring probe\n",
+			 NVSW_I2C_LAST_HOST_BUS_NR);
+		return -EPROBE_DEFER;
+	}
+	i2c_put_adapter(last_host_adapter);
+
+	/* Get I2C mux parent adapter */
+	nvsw_host_mux_i2c_adapter = i2c_get_adapter(NVSW_I2C_MUX_PARENT_BUS_NR);
+	if (!nvsw_host_mux_i2c_adapter) {
+		dev_err(&pdev->dev,
+			"I2C adapter for bus %d is not available\n",
+			NVSW_I2C_MUX_PARENT_BUS_NR);
+		return -ENODEV;
+	}
+
+	/* Create device at the top of host I2C tree */
+	nvsw_host_mux_brdinfo->platform_data = nvsw_host_mux_data;
+	nvsw_host_mux_i2c_client = i2c_new_client_device(nvsw_host_mux_i2c_adapter,
+						   nvsw_host_mux_brdinfo);
+	if (IS_ERR(nvsw_host_mux_i2c_client)) {
+		dev_err(&pdev->dev,
+			"Failed to create client %s at bus %d at addr 0x%02x\n",
+			nvsw_host_mux_brdinfo->type, NVSW_I2C_MUX_PARENT_BUS_NR,
+			nvsw_host_mux_brdinfo->addr);
+		err = PTR_ERR(nvsw_host_mux_i2c_client);
+		nvsw_host_mux_i2c_client = NULL;
+		goto i2c_new_client_device_fail;
+	}
+
+	nvsw_core = devm_kzalloc(&pdev->dev, sizeof(*nvsw_core), GFP_KERNEL);
+	if (!nvsw_core) {
+		err = -ENOMEM;
+		goto nvsw_host_probe_fail;
+	}
+
+	/* Set system configuration. */
+	nvsw_core->dev = &pdev->dev;
+	nvsw_core->hid = nvsw_host_hid;
+	nvsw_core->regmap_type = REGMAP_I2C;
+	nvsw_core->client = nvsw_host_mux_i2c_client;
+	nvsw_core->mux_num = mux_num;
+	for (i = 0; i < ARRAY_SIZE(nvsw_wd_data); i++)
+		nvsw_core->wd_data[i] = nvsw_wd_data[i];
+	nvsw_core->regio_data = nvsw_regs_io_data;
+	nvsw_core->led_data = nvsw_led_data;
+	nvsw_core->mux_init = nvsw_host_spc6_mux_topology_init;
+	nvsw_core->mux_exit = nvsw_host_spc6_mux_topology_exit;
+	platform_set_drvdata(pdev, nvsw_core);
+
+	err = nvsw_core_init(nvsw_core);
+	if (!err)
+		return 0;
+
+nvsw_host_probe_fail:
+	i2c_unregister_device(nvsw_host_mux_i2c_client);
+	nvsw_host_mux_i2c_client = NULL;
+i2c_new_client_device_fail:
+	i2c_put_adapter(nvsw_host_mux_i2c_adapter);
+	nvsw_host_mux_i2c_adapter = NULL;
+	return err;
+}
+
+static int nvsw_host_remove(struct platform_device *pdev)
+{
+	struct nvsw_core *nvsw_core = platform_get_drvdata(pdev);
+
+	nvsw_core_exit(nvsw_core);
+	i2c_unregister_device(nvsw_host_mux_i2c_client);
+	nvsw_host_mux_i2c_client = NULL;
+	i2c_put_adapter(nvsw_host_mux_i2c_adapter);
+	nvsw_host_mux_i2c_adapter = NULL;
+	return 0;
+}
+
+static struct platform_driver nvsw_host_driver = {
+	.driver		= {
+		.name	= NVSW_HOST_DEVICE_NAME,
+		.probe_type = PROBE_FORCE_SYNCHRONOUS,
+	},
+	.probe		= nvsw_host_probe,
+	.remove		= nvsw_host_remove,
+};
+
+static int __init nvsw_host_init(void)
+{
+	int err;
+
+	if (!dmi_check_system(nvsw_host_dmi_table))
+		return -ENODEV;
+
+	nvsw_host_pdev = platform_device_register_simple(NVSW_HOST_DEVICE_NAME,
+							 PLATFORM_DEVID_NONE,
+							 NULL, 0);
+	if (IS_ERR(nvsw_host_pdev))
+		return PTR_ERR(nvsw_host_pdev);
+
+	err = platform_driver_register(&nvsw_host_driver);
+	if (err)
+		platform_device_unregister(nvsw_host_pdev);
+
+	return err;
+}
+module_init(nvsw_host_init);
+
+static void __exit nvsw_host_exit(void)
+{
+	platform_driver_unregister(&nvsw_host_driver);
+	platform_device_unregister(nvsw_host_pdev);
+}
+module_exit(nvsw_host_exit);
+
+MODULE_AUTHOR("Felix Radensky <fradensky@nvidia.com>");
+MODULE_DESCRIPTION("Nvidia platform driver");
+MODULE_LICENSE("Dual BSD/GPL");
diff --git a/drivers/platform/mellanox/nvsw.h b/drivers/platform/mellanox/nvsw.h
new file mode 100644
index 0000000..5dd1c03
--- /dev/null
+++ b/drivers/platform/mellanox/nvsw.h
@@ -0,0 +1,300 @@
+/* SPDX-License-Identifier: GPL-2.0+
+ *
+ * Nvidia switch platform driver
+ *
+ * Copyright (C) 2025 Nvidia Technologies Ltd.
+ */
+
+#ifndef __NVSW_H__
+#define __NVSW_H__
+
+/* I2C bus IO offsets */
+#define NVSW_REG_CPLD1_VER_OFFSET	0x2500
+#define NVSW_REG_CPLD2_VER_OFFSET	0x2501
+#define NVSW_REG_CPLD3_VER_OFFSET	0x2502
+#define NVSW_REG_CPLD4_VER_OFFSET	0x2503
+#define NVSW_REG_CPLD1_PN_OFFSET	0x2504
+#define NVSW_REG_CPLD1_PN1_OFFSET	0x2505
+#define NVSW_REG_CPLD2_PN_OFFSET	0x2506
+#define NVSW_REG_CPLD2_PN1_OFFSET	0x2507
+#define NVSW_REG_CPLD3_PN_OFFSET	0x2508
+#define NVSW_REG_CPLD3_PN1_OFFSET	0x2509
+#define NVSW_REG_CPLD4_PN_OFFSET	0x250a
+#define NVSW_REG_CPLD4_PN1_OFFSET	0x250b
+#define NVSW_REG_CPLD7_VER_OFFSET	0x2510
+#define NVSW_REG_PG1_OFFSET		0x2511
+#define NVSW_REG_PG1_EVENT_OFFSET	0x2512
+#define NVSW_REG_PG1_MASK_OFFSET	0x2513
+#define NVSW_REG_PG2_OFFSET		0x2514
+#define NVSW_REG_PG2_EVENT_OFFSET	0x2515
+#define NVSW_REG_PG2_MASK_OFFSET	0x2516
+#define NVSW_REG_RESET_GP1_OFFSET	0x2517
+#define NVSW_REG_RESET_GP2_OFFSET	0x2519
+#define NVSW_REG_RESET_CAUSE_OFFSET	0x251d
+#define NVSW_REG_RESET_CAUSE1_OFFSET	0x251e
+#define NVSW_REG_RESET_CAUSE2_OFFSET	0x251f
+#define NVSW_REG_LED1_OFFSET		0x2520
+#define NVSW_REG_LED5_OFFSET		0x2524
+#define NVSW_REG_LED6_OFFSET		0x2525
+#define NVSW_REG_LED7_OFFSET		0x2526
+#define NVSW_REG_CPLD7_PN_OFFSET	0x2528
+#define NVSW_REG_CPLD7_PN1_OFFSET	0x2529
+#define NVSW_REG_GP0_RO_OFFSET		0x252a
+#define NVSW_REG_GP2_RO_OFFSET		0x252b
+#define NVSW_REG_GP1_RO_OFFSET		0x252c
+#define NVSW_REG_GPCOM0_OFFSET		0x252d
+#define NVSW_REG_GP0_OFFSET		0x252e
+#define NVSW_REG_GP7_OFFSET		0x252f
+#define NVSW_REG_GP1_OFFSET		0x2530
+#define NVSW_REG_GP4_OFFSET		0x2532
+#define NVSW_REG_GP5_RO_OFFSET		0x2533
+#define NVSW_REG_FIELD_UPGRADE		0x2534
+#define NVSW_REG_SAFE_BIOS_OFFSET	0x2535
+#define NVSW_REG_PWM_CONTROL_OFFSET	0x2537
+#define NVSW_REG_GP5_OFFSET		0x2538
+#define NVSW_REG_GP6_OFFSET		0x2539
+#define NVSW_REG_AGGR_OFFSET		0x253a
+#define NVSW_REG_AGGR_MASK_OFFSET	0x253b
+#define NVSW_REG_FU_CAP_OFFSET		0x253c
+#define NVSW_REG_BRD4_OFFSET		0x253d
+#define NVSW_REG_BRD4_EVENT_OFFSET	0x253e
+#define NVSW_REG_BRD4_MASK_OFFSET	0x253f
+#define NVSW_REG_AGGRLO_OFFSET		0x2540
+#define NVSW_REG_AGGRLO_MASK_OFFSET	0x2541
+#define NVSW_REG_AGGRCO_OFFSET		0x2542
+#define NVSW_REG_AGGRCO_MASK_OFFSET	0x2543
+#define NVSW_REG_BRD1_OFFSET		0x2547
+#define NVSW_REG_BRD1_EVENT_OFFSET	0x2548
+#define NVSW_REG_BRD1_MASK_OFFSET	0x2549
+#define NVSW_REG_HEALTH_OFFSET		0x254d
+#define NVSW_REG_HEALTH_EVENT_OFFSET	0x254e
+#define NVSW_REG_HEALTH_MASK_OFFSET	0x254f
+#define NVSW_REG_ASIC1_HEALTH_OFFSET	0x2550
+#define NVSW_REG_ASIC1_EVENT_OFFSET	0x2551
+#define NVSW_REG_ASIC1_MASK_OFFSET	0x2552
+#define NVSW_REG_ASIC2_HEALTH_OFFSET	0x2553
+#define NVSW_REG_ASIC2_EVENT_OFFSET	0x2554
+#define NVSW_REG_ASIC2_MASK_OFFSET	0x2555
+#define NVSW_REG_VR1_ALERT_OFFSET	0x2558
+#define NVSW_REG_VR1_ALERT_EVENT_OFFSET	0x2559
+#define NVSW_REG_VR1_ALERT_MASK_OFFSET	0x255a
+#define NVSW_REG_VR2_ALERT_OFFSET	0x255e
+#define NVSW_REG_VR2_ALERT_EVENT_OFFSET	0x255f
+#define NVSW_REG_VR2_ALERT_MASK_OFFSET	0x2560
+#define NVSW_REG_PS_DC_OK_OFFSET		0x2567
+#define NVSW_REG_PS_DC_OK_EVENT_OFFSET	0x2568
+#define NVSW_REG_PS_DC_OK_MASK_OFFSET	0x2569
+#define NVSW_REG_CPLD8_PN_OFFSET	0x2573
+#define NVSW_REG_CPLD8_PN1_OFFSET	0x2574
+#define NVSW_REG_PG3_OFFSET		0x2576
+#define NVSW_REG_PG3_EVENT_OFFSET	0x2577
+#define NVSW_REG_PG3_MASK_OFFSET	0x2578
+#define NVSW_REG_PG4_OFFSET		0x2579
+#define NVSW_REG_PG4_EVENT_OFFSET	0x257a
+#define NVSW_REG_PG4_MASK_OFFSET	0x257b
+#define NVSW_REG_CPLD6_VER_OFFSET	0x257c
+#define NVSW_REG_CPLD6_PN_OFFSET	0x257d
+#define NVSW_REG_CPLD6_PN1_OFFSET	0x257e
+#define NVSW_REG_CPLD10_PN_OFFSET	0x2580
+#define NVSW_REG_CPLD10_PN1_OFFSET	0x2581
+#define NVSW_REG_ASIC3_HEALTH_OFFSET	0x2582
+#define NVSW_REG_ASIC3_EVENT_OFFSET	0x2583
+#define NVSW_REG_ASIC3_MASK_OFFSET	0x2584
+#define NVSW_REG_FAN_OFFSET		0x2588
+#define NVSW_REG_FAN_EVENT_OFFSET	0x2589
+#define NVSW_REG_FAN_MASK_OFFSET	0x258a
+#define NVSW_REG_CPLD5_VER_OFFSET	0x258e
+#define NVSW_REG_CPLD5_PN_OFFSET	0x258f
+#define NVSW_REG_CPLD5_PN1_OFFSET	0x2590
+#define NVSW_REG_EROT_OFFSET		0x2591
+#define NVSW_REG_EROT_EVENT_OFFSET	0x2592
+#define NVSW_REG_EROT_MASK_OFFSET	0x2593
+#define NVSW_REG_EROT_ERR_OFFSET	0x2594
+#define NVSW_REG_EROT_ERR_EVENT_OFFSET	0x2595
+#define NVSW_REG_EROT_ERR_MASK_OFFSET	0x2596
+#define NVSW_REG_PWRB_OFFSET		0x2597
+#define NVSW_REG_PWRB_EVENT_OFFSET	0x2598
+#define NVSW_REG_PWRB_MASK_OFFSET	0x2599
+#define NVSW_REG_FRU1_OFFSET		0x25ac
+#define NVSW_REG_FRU1_EVENT_OFFSET	0x25ad
+#define NVSW_REG_FRU1_MASK_OFFSET	0x25ae
+#define NVSW_REG_LEAK_OFFSET		0x25af
+#define NVSW_REG_LEAK_EVENT_OFFSET	0x25b0
+#define NVSW_REG_LEAK_MASK_OFFSET	0x25b1
+#define NVSW_REG_GP4_RO_OFFSET		0x25c2
+#define NVSW_REG_SPI_CHNL_SELECT	0x25c3
+#define NVSW_REG_CPLD5_MVER_OFFSET	0x25c4
+#define NVSW_REG_CPLD8_VER_OFFSET	0x25c8
+#define NVSW_REG_CPLD9_VER_OFFSET	0x25c9
+#define NVSW_REG_CPLD10_VER_OFFSET	0x25ca
+#define NVSW_REG_CPLD9_PN_OFFSET	0x25cb
+#define NVSW_REG_CPLD9_PN1_OFFSET	0x25cc
+#define NVSW_REG_WD2_TMR_OFFSET		0x25cd
+#define NVSW_REG_WD2_TLEFT_OFFSET	0x25ce
+#define NVSW_REG_WD2_ACT_OFFSET		0x25cf
+#define NVSW_REG_WD3_TMR_OFFSET		0x25d1
+#define NVSW_REG_WD3_TLEFT_OFFSET	0x25d2
+#define NVSW_REG_WD3_ACT_OFFSET		0x25d3
+#define NVSW_REG_CPLD7_MVER_OFFSET	0x25d5
+#define NVSW_REG_CPLD8_MVER_OFFSET	0x25d6
+#define NVSW_REG_CPLD9_MVER_OFFSET	0x25d7
+#define NVSW_REG_CPLD10_MVER_OFFSET	0x25d8
+#define NVSW_REG_CPLD6_MVER_OFFSET	0x25d9
+#define NVSW_REG_MUX0_OFFSET		0x25da
+#define NVSW_REG_MUX1_OFFSET		0x25db
+#define NVSW_REG_MUX2_OFFSET		0x25dc
+#define NVSW_REG_CPLD1_MVER_OFFSET	0x25de
+#define NVSW_REG_CPLD2_MVER_OFFSET	0x25df
+#define NVSW_REG_CPLD3_MVER_OFFSET	0x25e0
+#define NVSW_REG_CPLD4_MVER_OFFSET	0x25e1
+#define NVSW_REG_UFM_VERSION_OFFSET	0x25e2
+#define NVSW_REG_PWM1_OFFSET		0x25e3
+#define NVSW_REG_TACHO1_OFFSET		0x25e4
+#define NVSW_REG_TACHO2_OFFSET		0x25e5
+#define NVSW_REG_TACHO3_OFFSET		0x25e6
+#define NVSW_REG_TACHO4_OFFSET		0x25e7
+#define NVSW_REG_TACHO5_OFFSET		0x25e8
+#define NVSW_REG_TACHO6_OFFSET		0x25e9
+#define NVSW_REG_TACHO7_OFFSET		0x25eb
+#define NVSW_REG_TACHO8_OFFSET		0x25ec
+#define NVSW_REG_TACHO9_OFFSET		0x25ed
+#define NVSW_REG_TACHO10_OFFSET		0x25ee
+#define NVSW_REG_TACHO11_OFFSET		0x25ef
+#define NVSW_REG_TACHO12_OFFSET		0x25f0
+#define NVSW_REG_FAN_CAP1_OFFSET	0x25f5
+#define NVSW_REG_FAN_DRW_CAP_OFFSET	0x25f7
+#define NVSW_REG_TACHO_SPEED_OFFSET	0x25f8
+#define NVSW_REG_PSU_I2C_CAP_OFFSET	0x25f9
+#define NVSW_REG_CONFIG1_OFFSET		0x25fb
+#define NVSW_REG_CONFIG2_OFFSET		0x25fc
+#define NVSW_REG_CONFIG3_OFFSET		0x25fd
+#define NVSW_REG_MIN			0x2500
+#define NVSW_REG_MAX			0x26ff
+
+#define NVSW_AGGR_MASK_COMEX		BIT(0)
+#define NVSW_LOW_AGGR_MASK_I2C		BIT(6)
+#define NVSW_AGGR_MASK			BIT(2)
+#define NVSW_LOW_AGGR_MASK_LOW		0xe0
+#define NVSW_REG_FRU1_MASK		GENMASK(3, 0)
+#define NVSW_LEAK_MASK			GENMASK(5, 0)
+#define NVSW_SSD_I2C_ALERT_MASK		BIT(0)
+#define NVSW_WD_EXP_MASK		BIT(1)
+#define NVSW_5V_USB_MASK		BIT(2)
+#define NVSW_PCB_TEMP1_MASK		BIT(3)
+#define NVSW_PCB_TEMP2_MASK		BIT(4)
+#define NVSW_SGMII_MASK			BIT(5)
+#define NVSW_SDD_PG_MASK		BIT(6)
+#define NVSW_LEAK_AGGR_MASK		BIT(7)
+#define NVSW_EROT_MASK			GENMASK(2, 0)
+#define NVSW_FAN_NG_MASK		GENMASK(6, 0)
+#define NVSW_ASIC_MASK			GENMASK(1, 0)
+#define NVSW_ASIC2_MASK			GENMASK(3, 2)
+#define NVSW_ASIC3_MASK			GENMASK(5, 4)
+#define NVSW_ASIC4_MASK			GENMASK(7, 6)
+#define NVSW_ASICS_MASK			GENMASK(7, 0)
+#define NVSW_LOW_AGGR_MASK_ASIC1	BIT(0)
+#define NVSW_LOW_AGGR_MASK_ASIC2	BIT(1)
+#define NVSW_LOW_AGGR_MASK_ASIC3	BIT(2)
+#define NVSW_PWR_BUTTON_MASK		BIT(0)
+#define NVSW_AMB_TEMP_SENSE_MASK	BIT(1)
+#define NVSW_GRACEFUL_POWER_OFF_MASK	BIT(2)
+#define NVSW_CPU_POWER_OFF_READY_MASK	BIT(3)
+#define NVSW_CPU_RESET_MASK		BIT(4)
+#define NVSW_APML_SMB_ALERT_MASK	BIT(5)
+#define NVSW_CPU_UNEXP_POWER_OFF_MASK	BIT(6)
+#define NVSW_UID_PUSH_BUTTON_MASK	BIT(7)
+#define NVSW_RTC_MASK			BIT(0)
+#define NVSW_HOT_SWAP_ALERT_MASK	BIT(1)
+#define NVSW_LED_LO_NIBBLE_MASK		GENMASK(7, 4)
+#define NVSW_LED_HI_NIBBLE_MASK		GENMASK(3, 0)
+#define NVSW_UART_SEL_MASK		GENMASK(7, 6)
+#define NVSW_BIOS_STATUS_MASK		GENMASK(3, 1)
+#define NVSW_REG_RESET_MASK		BIT(1)
+#define NVSW_MASTER_MASK		BIT(5)
+#define NVSW_FU_CAP_MASK		GENMASK(1, 0)
+#define NVSW_WD_RESET_ACT_MASK		GENMASK(7, 1)
+#define NVSW_WD_FAN_ACT_MASK		(GENMASK(7, 0) & ~BIT(4))
+#define NVSW_WD_TYPE2_TO_MASK		0
+#define NVSW_WD3_DFLT_TIMEOUT		600
+#define NVSW_I2C_CAP_BIT		0x04
+#define NVSW_I2C_CAP_MASK		GENMASK(5, NVSW_I2C_CAP_BIT)
+
+#define NVSW_NR_NONE			-1
+#define NVSW_MUX_MAX			2
+#define NVSW_WD_MAX			2
+
+ /* The system type. */
+enum nvsw_core_hid_type {
+	HID162,
+	HID176,
+	HID177,
+	HID180,
+	HID181,
+	HID182,
+	HID183,
+	HID185,
+	HID187,
+	HID188,
+	HID193,
+};
+
+ /* The system register map type. */
+enum nvsw_core_regmap_type {
+	REGMAP_IO,
+	REGMAP_I2C,
+};
+
+/* nvsw_core - device private data
+ * hid: hardware Id;
+ * regmap_type: register map type;
+ * @device: device;
+ * @client: I2C client device;
+ * @np: device node;
+ * @regmap: device register map;
+ * @regio_data: register access platform data;
+ * @regio: register access device;
+ * @hotplug_data: hotplug platform data;
+ * @hotplug: hotplug device;
+ * @led: led devices;
+ * @led_data: led platform data;
+ * @fan - fan device;
+ * @fan_data - fan platform data;
+ * @wd - watchdog device;
+ * @wd_data - watchdog platform data;
+ * @port_map - io port mapping;
+ * @set_reg_default - set default registers callback;
+ * @mux_init: mux initialization callback;
+ * @mux_exit: mux de-initialization callback;
+ * @mux: mux devices;
+ * @mux_num: number of mux device;
+ */
+struct nvsw_core {
+	enum nvsw_core_hid_type hid;
+	enum nvsw_core_regmap_type regmap_type;
+	struct device *dev;
+	struct i2c_client *client;
+	struct device_node *np;
+	struct regmap *regmap;
+	struct mlxreg_core_platform_data *regio_data;
+	struct platform_device *regio;
+	struct mlxreg_core_hotplug_platform_data *hotplug_data;
+	struct platform_device *hotplug;
+	struct platform_device *led;
+	struct mlxreg_core_platform_data *led_data;
+	struct platform_device *fan;
+	struct mlxreg_core_platform_data *fan_data;
+	struct platform_device *wd[NVSW_WD_MAX];
+	struct mlxreg_core_platform_data *wd_data[NVSW_WD_MAX];
+	void __iomem *(*port_map)(struct nvsw_core *nvsw_core);
+	int (*set_reg_default)(struct regmap *regmap);
+	int (*mux_init)(struct nvsw_core *nvsw_core);
+	void (*mux_exit)(struct nvsw_core *nvsw_core);
+	struct platform_device *mux[NVSW_MUX_MAX];
+	u32 mux_num;
+};
+
+/* Platform core init/exit functions. */
+int nvsw_core_init(struct nvsw_core *nvsw_core);
+void nvsw_core_exit(struct nvsw_core *nvsw_core);
+
+#endif /* !defined(__NVSW_H__) */
-- 
2.34.1

